### PR TITLE
treat fatal errors as actually fatal

### PR DESCRIPTION
**Pull request by bparees**
fixes https://github.com/openshift/origin/issues/6388

### CODE REVIEW

**Comment by bparees** [comment]
@jim-minter ptal.  among other things, we were not recognizing fatal errors as being fatal because they were pointers and our check only looked for non-pointer types.   Also am now treating Forbidden as a fatal error.  (we get a fatal error when the namespace is being deleted and we try to instantiate into it).

**Comment by bparees** [comment]
[test]

**Comment by bparees** [comment]
[testextended][extended:core(builds)]

File: pkg/build/controller/config_controller.go [hunk 1]
```diff
@@ -28,8 +28,12 @@ func (e ConfigControllerFatalError) Error() string {
 
 // IsFatal returns true if err is a fatal error
 func IsFatal(err error) bool {
-	_, isFatal := err.(ConfigControllerFatalError)
-	return isFatal
+	switch err.(type) {
+	case ConfigControllerFatalError, *ConfigControllerFatalError:
```
Comment (jim-minter) [inline pkg/build/controller/config_controller.go]: I think this is bad news - is it possible to fix the one place where we return &ConfigControllerFatalError (as opposed to ConfigControllerFatalError) instead?

File: pkg/build/controller/strategy/util.go [hunk 2]
```diff
@@ -37,8 +37,12 @@ func (e FatalError) Error() string {
 
 // IsFatal returns true if the error is fatal
 func IsFatal(err error) bool {
-	_, isFatal := err.(FatalError)
-	return isFatal
+	switch err.(type) {
+	case FatalError, *FatalError:
```
Comment (jim-minter) [inline pkg/build/controller/strategy/util.go]: Is there anywhere in the codebase that returns *FatalError?

File: pkg/build/generator/generator.go [hunk 3]
```diff
@@ -38,8 +38,12 @@ func (e GeneratorFatalError) Error() string {
 
 // IsFatal returns true if err is a fatal error
 func IsFatal(err error) bool {
-	_, isFatal := err.(GeneratorFatalError)
-	return isFatal
+	switch err.(type) {
+	case GeneratorFatalError, *GeneratorFatalError:
```
Comment (jim-minter) [inline pkg/build/generator/generator.go]: I think there's only one &GeneratorFatalError too

File: pkg/build/controller/config_controller.go (hunk 1 above)
Comment (bparees) [inline pkg/build/controller/config_controller.go]: I considered it, but nothing is going to stop someone else from going it in the future, so i'd want to check both regardless.

File: pkg/build/controller/strategy/util.go (hunk 2 above)
Comment (bparees) [inline pkg/build/controller/strategy/util.go]: not today, but again, who knows what someone does tomorrow.

**Comment by bparees** [comment]
@jim-minter I agree that in theory we can fix the code so it all works with non-pointer types, but I wanted to be defensive against what someone does in the future... do you see an issue w/ checking for both types, even if we aren't using both today?

**Comment by jim-minter** [comment]
OK, I stand by my instinct but realise my reasoning is wrong :-)

I think the actual issue is that we're not using a pointer receiver on `func (e ConfigControllerFatalError) Error() string` when we should be.  Ditto FatalError and GeneratorFatalError.

If these lines are re-written to `func (e *ConfigControllerFatalError) Error() string` (etc.), then only *ConfigControllerFatalError (etc.) can implement error, and it is no longer necessary to test for ConfigControllerFatalError (etc.).  I think that's the route forward.

**Comment by bparees** [comment]
this is one of the reasons i loathe golang's "it's sometimes a pointer" approach to pointers.

**Comment by bparees** [comment]
@jim-minter updated, ptal

**Comment by jim-minter** [comment]
@bparees lgtm

**Comment by bparees** [comment]
[merge]

**Pull request merged**